### PR TITLE
gsoc: rewrite compiler projects page to reflect current needs

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

File: soc/projects/compiler.md [hunk 1]
```diff
@@ -7,44 +7,49 @@ title:  Compiler Projects – Summer of Code
 
 {% include toc.html %}
 
-## C Linter
+## Improved code-generation for un-inferred code
 
-Memory errors in Julia's underlying C code are sometimes difficult to trace, and missing garbage-collector "roots" (GC roots) can lead to segfaults and other problems. One potential way to make it easier to find such errors would be to write a package that checks Julia's `src/` directory for missing GC roots. A Julia-based solution might leverage the [Clang.jl](https://github.com/ihnorton/Clang.jl) package to parse the C code, determine which call chains can trigger garbage collection, and then look for objects that lack GC root protection. Alternatively, the same strategy might be implemented in C++ by writing a plugin for [Clang's static analyzer](http://clang-analyzer.llvm.org/). Another attractive approach is to leverage [coccinelle](http://coccinelle.lip6.fr/).
+Currently the quality (perfomance) of generated code does not degrade particularly gracefully as the quality of type-inference degrades. There are a number of difference projects possible here. Some example optimization opportunities include:
```
Comment (KristofferC) [inline soc/projects/compiler.md]: `perfomance -> performance`

Comment (Sacha0) [inline soc/projects/compiler.md]: `difference` -> `different`

File: soc/projects/compiler.md [hunk 2]
```diff
@@ -7,44 +7,49 @@ title:  Compiler Projects – Summer of Code
...
 
-**Expected Results**: A tool that, when run against Julia's `src/` directory, highlights lines that need to have additional GC roots added.
+- Calling convention for varargs functions (https://github.com/JuliaLang/julia/issues/5402)
+- Call-site splatting optimizations (https://github.com/JuliaLang/julia/issues/13359)
+- Keyword argument inference / inlining
+- More calling conventions, permitting improved performance (unboxed argument passing) when the direct method target is not known
+- Value domination and liveness analysis for optimizing stack space and gc-root allocation
 
-**Required Skills**: Familiarity with C, runtimes, and experience with (or interest in) static analysis.
+**Recommended Skills**: Experience with reading LLVM IR or assembly. An understanding of formal logic and using it develop improved code transformations.
```
Comment (Sacha0) [inline soc/projects/compiler.md]: `using it develop` -> `using it to develop` or perhaps `and its application to improving code transformations`?

File: soc/projects/compiler.md [hunk 3]
```diff
@@ -7,44 +7,49 @@ title:  Compiler Projects – Summer of Code
...
 
-## Specialized call-site method caching
+The starting point should likely be https://github.com/JuliaLang/Juleps/pull/21. This will need further elaboration and study to identify any areas that are likely to be problematic under the new approach. Also will require new code and testing to demonstrate the performance and usability of the new code patterns; and to begin deprecation and replacement of the existing code pattern with Nullables.
 
-Julia's method cache is shared by all call sites of a generic function. Although whenever single dispatch is provable we generate a direct call, there are some cases where dynamic dispatch is inevitable. When the compiler can prove (using type inference) that the possible matches for a call site is small enough, it would be a huge performance win to generate a small cache specific to this call site.
+**Expected Results**: Initiate replacement of the `Nullable{T}` with an implementation based around `Union{T, Void}` and `Union{Some{T}, Void}`.
 
-**Expected Results**: A PR for Base Julia implementing this optimisation.
+**Recommended Skills**: Experience with performance testing and Julia. Experience with data analysis also a big advantage.
```
Comment (Sacha0) [inline soc/projects/compiler.md]: `data analysis` -> `data analysis involving missing values`?

**Comment by shashi** [comment]
Thanks @vtjnash @KristofferC and @Sacha0! I made those changes.

**Comment by vtjnash** [comment]
Thanks @shashi

**Comment by timholy** [comment]
@vtjnash, speaking of compiler projects, is there any support for adding "finish https://github.com/JuliaLang/julia/pull/18632" as a GSoC project? Or is it beyond the ken of all mortals besides @carnaval?

**Comment by vtjnash** [comment]
I would definitely support that. It could be quite difficult finding a student capable of taking on that project though, especially in one summer (the scope of it is potentially quite large for nailing down semantics).

**Comment by shashi** [comment]
Would it make sense to have just the stack allocation optimization from that PR as a GSoC project?

**Pull request merged**